### PR TITLE
Add RegisterLegalizationPasses() into the interface

### PR DESCRIPTION
**Pull request by antiagainst**
Addresses #968

### CODE REVIEW

**Comment by antiagainst** [comment]
This is the same recipe as used in glslang right now: https://github.com/KhronosGroup/glslang/blob/4036ef154fb0e11a2b235afce3e9b3b819cf5fea/SPIRV/GlslangToSpv.cpp#L6060-L6071

**Comment by antiagainst** [comment]
cc @greg-lunarg

**Comment by s-perron** [comment]
Do we have any ideas which ones are really for legalization and which ones for optimization?  This looks like the list that recipe that LunarG suggests to their customer and what we used for the performance and size optimization recipes.

File: include/spirv-tools/optimizer.hpp [hunk 1]
```diff
@@ -77,6 +77,11 @@ class Optimizer {
   // method.
   Optimizer& RegisterPass(PassToken&& pass);
 
+  // Registers passes that attempt to legalize the generated code.
+  // This sequence of passes is subject to constant review and will change
```
Comment (dneto0) [inline include/spirv-tools/optimizer.hpp]: It's worth saying these are passes used to legalize code originally generated by legacy HLSL.  Also say where to find a more current list, i.e. in Glslang

**Comment by antiagainst** [comment]
@s-perron: Legalization passes are largely general optimization passes since the transformations needed to legalize are similar: function inlining, load/store eliminating, const propagating and folding, DCE, etc. But legalization emphasizes transformations that flatten structs/arrays, which I'd assume performance/size optimization does not. Also legalization does not really care about loop unrolling and other performance oriented passes. So there are subtle differences.

Right now these recipes are almost identical. But they will drift away gradually I think.

**Comment by dnovillo** [comment]
I wish they were not referred to "optimizations", though.  I understand
that they are transformations required to produce canonical/legal vulkan.
Perhaps we can call them "legalizing transformations"?  I'm being picky,
but the term "optimization" is very specific.  It does not make much sense
to need an optimization for correctness.  They should be orthogonal.

OK, stepping off the soap box.

**Comment by antiagainst** [comment]
Haha, yes, I totally agree. You'll notice that in my previous reply I've used optimization/legalization/transformation carefully. ;-P

File: include/spirv-tools/optimizer.hpp (hunk 1 above)
Comment (antiagainst) [reply include/spirv-tools/optimizer.hpp]: Done.

**Comment by dneto0** [comment]
Rebased, squashed, and pushed into master as aec60b8158

**Pull request closed without merging**